gadget/mountedfilesystem.go

```diff
@@ -179,10 +198,578 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+//
+// The update process is composed of 2 main passes, and an optional rollback:
+//
+// 1) backup, where update data and current data is analyzed to identify
+// identical content, stamp files are created for entries that are to be
+// preserved, modified or otherwise touched by the update, that is, for existing
+// files that would be created/overwritten, ones that are explicitly listed as
+// preserved, or directories to be written to
+//
+// 2) update, where update data is written to the target location
+//
+// 3) rollback (optional), where update data is rolled back and replaced with
+// backup copies of files, newly created directories are removed
+type MountedFilesystemUpdater struct {
+	*MountedFilesystemWriter
+	backupDir   string
+	mountLookup mountLookupFunc
+}
+
+// NewMountedFilesystemUpdater returns an updater for given filesystem
+// structure, with structure content coming from provided root directory. The
+// mount is located by calling a mount lookup helper. The backup directory
+// contains backup state information for use during rollback.
+func NewMountedFilesystemUpdater(rootDir string, ps *PositionedStructure, backupDir string, mountLookup mountLookupFunc) (*MountedFilesystemUpdater, error) {
+	fw, err := NewMountedFilesystemWriter(rootDir, ps)
+	if err != nil {
+		return nil, err
+	}
+	if mountLookup == nil {
+		return nil, fmt.Errorf("internal error: mount lookup helper must be provided")
+	}
+	if backupDir == "" {
+		return nil, fmt.Errorf("internal error: backup directory must not be unset")
+	}
+	fu := &MountedFilesystemUpdater{
+		MountedFilesystemWriter: fw,
+		backupDir:               backupDir,
+		mountLookup:             mountLookup,
+	}
+	return fu, nil
+}
+
+func fsStructBackupPath(backupDir string, ps *PositionedStructure) string {
+	return filepath.Join(backupDir, fmt.Sprintf("struct-%v", ps.Index))
+}
+
+// entryDestPaths resolves destination and backup paths for given
+// source/target combination. Backup location is within provided
+// backup directory or empty if directory was not provided.
+func (f *MountedFilesystemUpdater) entryDestPaths(dstRoot, source, target, backupDir string) (dstPath, backupPath string) {
+	dstBasePath := target
+	if strings.HasSuffix(target, "/") {
+		// write to a directory
+		dstBasePath = filepath.Join(dstBasePath, filepath.Base(source))
+	}
+	dstPath = filepath.Join(dstRoot, dstBasePath)
+
+	if backupDir != "" {
+		backupPath = filepath.Join(backupDir, dstBasePath)
+	}
+
+	return dstPath, backupPath
+}
+
+// entryPaths resolves the source, destination and backup paths for given
+// source/target combination. The source location is always within the root
+// directory passed during initialization. Backup location is within provided
+// backup directory or empty if directory was not provided.
+func (f *MountedFilesystemUpdater) entryPaths(dstRoot, source, target, backupDir string) (srcPath, dstPath, backupPath string) {
```

Comment:
nitpick: this is not called so frequently, maybe the couple of places should just call the separate helpers

Reply:
Dropped entryPaths()